### ISSUE TITLE
Sendmail truncates transcoded address field values

### ISSUE THREAD

**Issue by Vitaliy-1** [pkp/pkp-lib#7239]
**Describe the bug**
Field values encoded with `imap_8bit` can be truncated if fall under a certain length. Relevant for `From` and `Reply-To` header fields. 

**To Reproduce**
The problem appears when sending with integrated Laravel Mail Service through Swift Mailer and Sendmail.
I've tried to reproduce it directly via Sendmail and the result is the same:
1. Pick relatively long Cyrillic full name, e.g. `Михайло Сергійович Грушевський` and encode it with Swift Mailer approach, which is `imap_8bit` encoding with a special type of metacharacters and line breaks:
```
=?utf-8?Q?=D0=9C=D0=B8=D1=85=D0=B0=D0=B9=D0=BB=D0=BE_?=
 =?utf-8?Q?=D0=A1=D0=B5=D1=80=D0=B3=D1=96=D0=B9=D0=BE=D0=B2?=
 =?utf-8?Q?=D0=B8=D1=87_=D0=93=D1=80=D1=83=D1=88=D0=B5?=
 =?utf-8?Q?=D0=B2=D1=81=D1=8C=D0=BA=D0=B8=D0=B9?=
```
2. Create a text file with the following content
```
To: john.doe@anyemailaddress.example
From: =?utf-8?Q?=D0=9C=D0=B8=D1=85=D0=B0=D0=B9=D0=BB=D0=BE_?=
 =?utf-8?Q?=D0=A1=D0=B5=D1=80=D0=B3=D1=96=D0=B9=D0=BE=D0=B2?=
 =?utf-8?Q?=D0=B8=D1=87_=D0=93=D1=80=D1=83=D1=88=D0=B5?=
 =?utf-8?Q?=D0=B2=D1=81=D1=8C=D0=BA=D0=B8=D0=B9?=
 <anyuser@anyemailaddress.example>
```
`To` field should contain a valid email address 
3. Send it from a command line: `sendmail -vt < mail.txt`
4. The message in my case can't be sent (I've blocked email ports) and can be found in the queue in `/var/spool/mqueue`, where the `From` field is truncated:
```
H??From: =?utf-8?Q?=D0=9C=D0=B8=D1=85=D0=B0=D0=B9=D0=BB=D0=BE_?=
 =?utf-8?Q?=D0=A1=D0=B5=D1=80=D0=B3=D1=96=D0=B9=D0=BE=D0=B2?=
 =?utf-8?Q?=D0=B8=D1=87_=D0=93=D1=80=D1=83=D1=88=D0=B5?=
 =?utf-8?Q?=D0=B2=D1=81=D1=8C=D0=BA=D0=B8=D0=B9?=
 <anyuser@anyemailaddress.examp
.
```

**What application are you using?**
 Tested on Ubuntu, Sendmail v. 8.15.2.  

**Additional information**
I'm testing this locally, maybe this is the cause. Nevertheless, think it's better to fill this issue and debug this in more detail before the release.

**Comment by Vitaliy-1** [comment]
After switching to Symfony Mailer, I'm still able to reproduce the issue. I'd say that sendmail limits email address to 256 symbols and it isn't handled in any way by the Mailer

**Comment by Vitaliy-1** [comment]
Regarding `To` header field value, it affects each user (name + address)  in the list

**Comment by Vitaliy-1** [comment]
To summarise findings: 
1. It affects sendmail but not SMTP transport (configured and tested both).
2. Looks more like an internal sendmail issue.
3. Despite the problem, email still can be delivered to the specified address (e.g., if `To` header field body is truncated). The received email indeed contains an invalid address field value as shown above. 
4. In short, the email address gets truncated if the total number of symbols count exceeds 210; not counting symbols, which embrace new line character sequence: `=?utf-8?Q?` at the start and `?=` at the end of a new line but including `<...>`
5. Sendmail removes the name if it has 205 encoded symbols or more, including `"..."`. And it should do this depending on the total symbols count instead.

Looking at Symfony email address abstraction, we can do sendmail's work in terms of controlling the total number of symbols in the address header field body just by performing the check using `Symfony\Component\Mime\Address::toString()` additionally encoded with `quoted_printable_encode()`;

**Comment by Vitaliy-1** [comment]
PRs
pkp-lib: https://github.com/pkp/pkp-lib/pull/8377
tests: https://github.com/pkp/ojs/pull/3590

**Comment by Vitaliy-1** [comment]
@asmecher, can you take a look at this PR?

**Comment by asmecher** [comment]
@Vitaliy-1, I've done a bit of local testing in trying to reproduce the problem. Sending directly via SMTP to Mailhog seems to work fine:
![image](https://user-images.githubusercontent.com/200411/200901105-8c2134a2-a96c-4203-9683-cfb72cb7948b.png)

I configured Sendmail locally (as best I can from `localhost`) and tried to replicate the truncation you experienced. Here is my test PHP source:
```php
<?php

use Symfony\Component\Mailer\Mailer;
use Symfony\Component\Mailer\Transport\SendmailTransport;
use Symfony\Component\Mailer\Transport;
use Symfony\Component\Mime\Email;

require_once('lib/pkp/lib/vendor/autoload.php');

$transport = new SendmailTransport();
//$transport = Transport::fromDsn('smtp://symfony-mailer@localhost:1025');

$mailer = new Mailer($transport);

$email = (new Email())
    ->from('asmecher@localhost')
    ->to('Михайло Сергійович Грушевський <asmecher@localhost>')
    ->priority(Email::PRIORITY_HIGHEST)
    ->subject('My first mail using Symfony Mailer')
    ->text('This is an important message!')
    ->html('<strong>This is an important message!</strong>');

$mailer->send($email);
```
Running this drops the following into `/var/spool/mail`:
```
From asmecher@localhost  Wed Nov  9 09:34:39 2022
Return-Path: <asmecher@localhost>
Received: from localhost (localhost [127.0.0.1])
        by localhost (8.15.2/8.15.2/Debian-18) with ESMTP id 2A9HYdBk3778145
        for <asmecher@localhost>; Wed, 9 Nov 2022 09:34:39 -0800
Received: from [127.0.0.1] (asmecher@localhost)
        by friendo (8.15.2/8.15.2/Submit) with SMTP id 2A9HM0Ev3775587
        for <asmecher@localhost>; Wed, 9 Nov 2022 09:22:00 -0800
From: asmecher@localhost
To: =?utf-8?Q?=D0=9C=D0=B8=D1=85=D0=B0=D0=B9=D0=BB=D0=BE_?=
 =?utf-8?Q?=D0=A1=D0=B5=D1=80=D0=B3=D1=96=D0=B9=D0=BE=D0=B2?=
 =?utf-8?Q?=D0=B8=D1=87_=D0=93=D1=80=D1=83=D1=88=D0=B5?=
 =?utf-8?Q?=D0=B2=D1=81=D1=8C=D0=BA=D0=B8=D0=B9?= <asmecher@localhost>
X-Priority: 1 (Highest)
Subject: My first mail using Symfony Mailer
Message-ID: <f93c559999610856770d5f66073682bb@localhost>
MIME-Version: 1.0
Date: Wed, 09 Nov 2022 09:22:00 -0800
Content-Type: multipart/alternative; boundary=Wk7j3BuT

--Wk7j3BuT
Content-Type: text/plain; charset=utf-8
Content-Transfer-Encoding: quoted-printable

This is an important message!
--Wk7j3BuT
Content-Type: text/html; charset=utf-8
Content-Transfer-Encoding: quoted-printable

<strong>This is an important message!</strong>
--Wk7j3BuT--

```
...which doesn't appear to show truncation.

I also tried through my ISP's pipeline (LTMP/Exim) and that was received without truncation problems:
![image](https://user-images.githubusercontent.com/200411/200903171-d2cec03a-bafe-44e2-ae6d-844c47af4a48.png)

Long story short, I'm having trouble reproducing this in a variety of scenarios. Do you see anything that sticks out as a key difference?

**Comment by Vitaliy-1** [comment]
@asmecher, there are not enough symbols in the encoded `To` header body, e.g., in my example with this name, the email address contained 33 characters (including `<>`) and only 3 last ones were truncated

I used sendmail v 8.15.2 and sent the email to my own Gmail address to be sure that it was not a problem with logging.

**Comment by asmecher** [comment]
Thanks, that helps. The limit appears to be 256 bytes.

I was able to track this down to a [known Sendmail bug](https://github.com/Distrotech/sendmail/blob/master/KNOWNBUGS#L17..L21) (link is not to an official repo but the file is the same):
> * Header values which are too long may be truncated.
> 
>  If a value of a structured header is longer than 256 (MAXNAME)
>  characters then it may be truncated during output. For example,
>  if a single address in the To: header is longer than 256 characters
>  then it will be truncated which may result in a syntactically
>  invalid address.

In my own Sendmail config, I was still able to successfully send messages that exceeded this limit. When an address exceeded 256 characters, I got a delivery warning from Sendmail:
> sendmail[3794651]: 2A9NfX6X3794651: Dropped invalid comments from header address

...but despite this the message was still delivered with a valid "from" address. (It was delivered without the display name, which is the end result of that warning condition being hit.) I also tested this for overlong "to" addresses.

I [tracked this message down](https://github.com/Distrotech/sendmail/blob/master/sendmail/headers.c#L1635..L1642) in the Sendmail source to see whether the address is being preserved despite losing the display name. The message comes from a bail-out case for the [`crackaddr` function](https://github.com/Distrotech/sendmail/blob/master/sendmail/headers.c#L1249), which breaks the address (display name, email address, etc.) out of its mixed form into what Sendmail needs for delivery.

So I wonder whether your case, where you're looking in the mail queue, is processed differently than mine, where my Sendmail forwards the message along for delivery via SMTP. But either way, this is clearly a known Sendmail bug and I'd argue that we shouldn't be fixing it way up in a PHP application. I recommend we keep the PRs around in case anyone wants to use them, but I'm against merging it.

How we ever managed to build a viable global computer network in C is 100% beyond me! Rust can't replace all this macro expansion insanity fast enough.

**Comment by asmecher** [comment]
I'm going to close this as a Sendmail bug rather than something in our stack. But if it does rear its head again, and proves to be different from what's documented above, please note it here!
